java/client/src/main/java/glide/internal/AsyncRegistry.java

```diff
@@ -0,0 +1,334 @@
+    }
+
+    /**
+     * Complete callback with proper race condition handling. Supports both regular Java objects and
+     * DirectByteBuffer for large responses (>16KB).
+     */
+    public static boolean completeCallback(long correlationId, Object result) {
+        CompletableFuture<Object> future = activeFutures.get(correlationId);
+
+        if (future == null) {
+            // Future already completed or timed out
+            return false;
+        }
+
+        // complete() returns false if already completed
+        // This prevents IllegalStateException from completing twice
+        boolean completed = future.complete(result);
+
+        // Note: cleanup happens automatically in whenComplete()
+        // No manual removal needed, which eliminates race conditions
+
+        return completed;
+    }
+
+    /** Complete with error using same race-free pattern */
+    public static boolean completeCallbackWithError(long correlationId, String errorMessage) {
+        CompletableFuture<Object> future = activeFutures.get(correlationId);
+
+        if (future == null) {
+            return false;
+        }
+
+        RuntimeException exception = mapToTypedException(errorMessage);
+
+        // completeExceptionally() is also atomic and safe
+        boolean completed = future.completeExceptionally(exception);
+
+        // Note: cleanup happens automatically in whenComplete()
+        // No manual removal needed, which eliminates race conditions
+
+        return completed;
+    }
+
+    /**
+     * Complete with error using a structured error code from native layer. Codes map to glide-core
+     * RequestErrorType: 0-Unspecified, 1-ExecAbort, 2-Timeout, 3-Disconnect.
+     */
+    public static boolean completeCallbackWithErrorCode(
+            long correlationId, int errorTypeCode, String errorMessage) {
+        CompletableFuture<Object> future = activeFutures.get(correlationId);
+        if (future == null) {
+            return false;
+        }
+
+        String msg =
+                (errorMessage == null || errorMessage.isBlank())
+                        ? "Unknown error from native code"
+                        : errorMessage;
+        RuntimeException ex;
+        switch (errorTypeCode) {
+            case 3: // Disconnect
+                ex = new glide.api.models.exceptions.ClosingException(msg);
+                break;
+            case 0: // Unspecified - check if it's actually a connection error
+                // For unspecified errors, check the message to determine the actual type
+                ex = mapToTypedException(msg);
+                break;
+            case 1: // ExecAbort
+            case 2: // Timeout
+            default:
+                ex = new glide.api.models.exceptions.RequestException(msg);
+                break;
+        }
+
+        return future.completeExceptionally(ex);
+    }
+
+    private static RuntimeException mapToTypedException(String errorMessage) {
+        String msg =
```

Comment:
Can this be based on an error code at all? Using error message text is unreliable since localization can come into play.